File: Cloud Storage Samples/ShareFile/javascript/sharefile.js
/*
 * Cloud Storage 2022 JavaScript Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
 
const readline = require("readline");
const cloudstorage = require("@nsoftware/cloudstorage");

if(!cloudstorage) {
  console.error("Cannot find cloudstorage.");
  process.exit(1);
}
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

const path = require("path"); // Retrieves information about passed paths.
const fs = require("fs"); // Ensures that a provided file is valid.
let lastPrompt = ""; // The key for the last prompt sent to the user.
let lastDefault = ""; // The last default for the last prompt sent to the user.
let currentFolderId = ""; // The ID of the folder that the user is currently in.
let currentParentId = ""; // The ID of the parent to the folder that the user is currently in.
let subdomain = ""; // The account subdomain for the currently authenticated user.
let authStr = ""; // The auth string for the currently authenticated user.
const sharefile = new cloudstorage.sharefile();
const oauth = new cloudstorage.oauth();
main();


async function main(){


  console.log("\n*****************************************************************");
  console.log("\n* This demo shows how to use the ShareFile component to list,   *");
  console.log("\n* upload, download, delete, and share documents from ShareFile. *");
  console.log("\n* Default prompt values are in [ ] where applicable.            *");
  console.log("\n*****************************************************************\n");

    /*
    * NOTE: ShareFile requires a registered redirect URI which is https://
    *       and explicitly disallows localhost. A valid public domain with SSL enabled must
    *       be specific when creating your ClientId and ClientSecret values
    *    
    *       Visit https://api.sharefile.com/apikeys to create your credentials
          
    * IMPORTANT: 
    *       Because of the ShareFile restrictions detailed above this demo
    *       make use of an existing public URL (oauth.nsoftware.com)
    *       This should NOT be used in production. It is only used here to
    *       facilitate running of this demo. In a real application you should
    *       register your own Redirect URI.
    *      
    *       For any questions please contact support@nsoftware.com 
    */

  prompt("subdomain", "Account Subdomain", ":", "");
  rl.on('line', async (line) => {
    try{
      switch (lastPrompt) {
        // Enter account subdomain.
        case "subdomain":
          if(line === ""){
            subdomain = lastDefault;
          } else {
            subdomain = line;
          }
          prompt("client_id", "Client ID", ":", "Hb7kR4d6mBvQxnOe9Eg6V8CMIVjQ67vD");
        break;
        // Enter OAuth client ID.
        case "client_id":
          if(line === ""){
            oauth.setClientId(lastDefault);
          } else {
            oauth.setClientId(line);
          }
          prompt("client_secret", "Client Secret", ":", "jHAZH6f1Ac5O9krW9az3fLYFEZW4KoF1sr1hqDoVgrvv1zwc");
        break;
        // Enter OAuth client secret.
        case "client_secret" :
          if(line === ""){
            oauth.setClientSecret(lastDefault);
          } else {
            oauth.setClientSecret(line);
          }
          oauth.setClientProfile(cloudstorage.OauthClientProfiles.ocpApplication);
          oauth.setServerAuthURL("https://secure.sharefile.com/oauth/authorize");
          oauth.setServerTokenURL(`https://${subdomain}.sharefile.com/oauth/token`);
          
          oauth.setWebServerPort(7878); // Specify a particular port, as it will be used in the redirect.
          await oauth.startWebServer(); // In most cases, the webserver is automatically started when getAuthorization is called.  In this case, due to the use of oauth.nsoftware.com, it is started manually.
          oauth.setReturnURL("https://oauth.nsoftware.com/oauthdemo"); // This URL should NOT be used in a production application.  It is only for testing purposes.
          oauth.addParam("state", `http://localhost:${oauth.getWebServerPort()}`); // The oauth.nsoftware.com server uses this value to redirect to the local server embedded within the OAuth class.

          //oauth.setRefreshToken(""); // Set to value from previous runs to expedite the login process.
          authStr = await oauth.getAuthorization();
          //console.log(oauth.getRefreshToken()); // RefreshToken may be used to automatically obtain a new token without user interaction.
          await displayList("home");
          lastPrompt = "";
        break;
        // Retrieve number corresponding to the folder the user wishes to view.
        case "cd":
          let cd_choice = parseInt(line);
          if(cd_choice === NaN){
            console.log("Please choose a valid number corresponding to a folder.");
            lastPrompt = "";
            break;
          }
          if(cd_choice === -1){
            await displayList(currentParentId);
          } else {
            await displayList(sharefile.getItems().get(cd_choice).getId());
          }
          lastPrompt = "";
        break;
        // Retrieve download location.
        case "get_location" :
          sharefile.setLocalFile(line);
          prompt("get_file","Number corresponding to item you wish to download", ":", "" );
        break;
		// Retrieve number corresponding to the item the user wishes to download.
        case "get_file" :
          let get_choice = parseInt(line);
          if(get_choice === NaN){
            console.log("Please choose a valid number corresponding to an item.");
            lastPrompt = "";
            break;
          }
          if(sharefile.getItems().get(get_choice).getType() === cloudstorage.ShareFileItemTypes.sfitFile){
            sharefile.setLocalFile(sharefile.getLocalFile() + sharefile.getItems().get(get_choice).getName());
          } else if(sharefile.getItems().get(get_choice).getType() === cloudstorage.ShareFileItemTypes.sfitFolder){
            sharefile.setLocalFile(sharefile.getLocalFile() + sharefile.getItems().get(get_choice).getName() + ".zip");
          } else {
            console.log("Item is of an invalid type.");
            lastPrompt = "";
          }
          await sharefile.downloadFile(sharefile.getItems().get(get_choice).getId());
          lastPrompt = "";
        break;
		// Retrieve number corresponding to the file the user wishes to upload.
        case "put" :
          if(!fs.existsSync(line)){
            console.log("Please input valid file.");
            lastPrompt = "";
            break;
          }

          sharefile.setLocalFile(line);
          await sharefile.uploadFile(path.parse(line).base, currentFolderId);
          await displayList(currentFolderId);
          lastPrompt = "";
        break;
        // Retrieve number corresponding to the item the user wishes to delete.
        case "del" :
          let del_choice = parseInt(line)
          if (del_choice === NaN){
            console.log("Please choose a valid number corresponding to an item.");
            lastPrompt = "";
            break;
          }
          await sharefile.deleteItem(sharefile.getItems().get(del_choice).getId());
          await displayList(currentFolderId);
          lastPrompt = "";
        break;
        // Retrieve new folder name.
        case "mkdir" :
          await sharefile.createFolder(line, currentFolderId);
          await displayList(currentFolderId);
          lastPrompt = "";
        break;
        // Retrieve number corresponding to the item the user wishes to share.
        case "shr" :
          let shr_choice = parseInt(line)
          if (shr_choice === NaN){
            console.log("Please choose a valid number corresponding to an item.");
            lastPrompt = "";
            break;
          }
          console.log(`Share Link: ${await sharefile.createLink(sharefile.getItems().get(shr_choice).getId())}`);
          lastPrompt = "";
        break;
      }
      switch (line) {
        case "?":
          printMenu();
          menuPrompt();
        break;
        case "cd":
          prompt("cd", "Number corresponding to folder you wish to view (-1 to go up a level)", ":", "");
        break;
        case "ls": 
          await displayList(currentFolderId)
          printMenu();
          menuPrompt();       
        break;
        case "get" :
          prompt("get_location", "Directory in which to download", ":", "");
        break;
        case "put" :
          prompt("put", "Path of file that will be uploaded", ":", "");
        break;
        case "del" :
          prompt("del", "Number corresponding to item you wish to delete", ":", "");
        break;
        case "mkdir" :
          prompt("mkdir", "Name of new folder", ":", "");
        break;
        case "shr" :
          prompt("shr", "Number corresponding to item you wish to share", ":", "");
        break;
        case "req" :
          console.log(`Request Link: ${await sharefile.createRequestLink(currentFolderId)}`);
          printMenu();
          menuPrompt();
        break;
        case "q":
          process.exit();
        break;
        default:
        if (lastPrompt === ""){
          printMenu();
          menuPrompt();
        }
        break;
      }
    } catch (error){
      console.log(error.message);      
      process.exit(1);
    }
  });

}

async function displayList(folderId){
  sharefile.setAccountSubdomain(subdomain);
  sharefile.setAuthorization(authStr);

  await sharefile.getItemInfo(folderId);
  currentFolderId = sharefile.getItems().get(0).getId();
  currentParentId = sharefile.getItems().get(0).getParentId();
  
  console.log("\n",  `#  |  Title  |  Type  |  Creation Date`);
  console.log( `-1 |  ..  |  Folder  |  ..`);

  await sharefile.listItems(currentFolderId);
  let i = 0;
  for(const item of sharefile.getItems()){
    console.log(`${i}  |  ${item.getName()}  |  ${resolveItemType(item.getType())}  |  ${item.getCreationDate()}`);
    i = i + 1;
  }
}

function resolveItemType(type){
  switch (type){
    case cloudstorage.ShareFileItemTypes.sfitFile:
      return "File";
    case cloudstorage.ShareFileItemTypes.sfitFolder:
      return "Folder";
    case cloudstorage.ShareFileItemTypes.sfitLink:
      return "Link";
    case cloudstorage.ShareFileItemTypes.sfitNote:
      return "Note";
    case cloudstorage.ShareFileItemTypes.sfitSymbolicLink:
      return "Symbolic Link";
    default:
      return "";
  }
}

function printMenu() {
  console.log("\n?     -   Help");
  console.log("cd    -   Change Folders");
  console.log("ls    -   List Items");
  console.log("get   -   Download Item");
  console.log("put   -   Upload File");
  console.log("del   -   Delete Item");
  console.log("mkdir -   Create Folder");
  console.log("shr   -   Create Share Link");
  console.log("req   -   Create Request Link");
  console.log("q     -   Quit\n");	
  lastPrompt = "";
}
function menuPrompt(){
  process.stdout.write('>: ');
  lastPrompt = "";  
}


function prompt(promptName, label, punctuation, defaultVal)
{
  lastPrompt = promptName;
  lastDefault = defaultVal;
  process.stdout.write(`${label} [${defaultVal}] ${punctuation} `);
}

File: Cloud Storage Samples/ShareFile/python/sharefile.py
# 
# Cloud Storage 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of Cloud Storage in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/cloudstorage
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from cloudstorage import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)




#Global Variables#
currentFolderId = ""
currentParentId = ""
subdomain = ""
authStr = ""
sharefile = ShareFile()

#  NOTE: ShareFile requires a registered redirect URI which is https://
#        and explicitly disallows localhost. A valid public domain with SSL enabled must
#        be specified when creating your ClientId and ClientSecret values.
#        
#        Visit https://api.sharefile.com/apikeys to create your credentials.
#        
#  IMPORTANT: 
#       Because of the ShareFile restrictions detailed above this demo
#       make use of an existing public URL (oauth.nsoftware.com)
#       This should NOT be used in production. It is only used here to
#       facilitate running of this demo. In a real application you should
#       register your own Redirect URI.
#       
#       For any questions please contact support@nsoftware.com.
def authenticate():
  global authStr, subdomain

  oauth = OAuth()
  oauth.set_client_profile(0) #Application
  oauth.set_client_id(prompt("Client ID", "Hb7kR4d6mBvQxnOe9Eg6V8CMIVjQ67vD"))
  oauth.set_client_secret(prompt("Client Secret", "jHAZH6f1Ac5O9krW9az3fLYFEZW4KoF1sr1hqDoVgrvv1zwc"))
  oauth.set_server_auth_url("https://secure.sharefile.com/oauth/authorize")
  oauth.set_server_token_url("https://" + subdomain + ".sharefile.com/oauth/token")

  oauth.set_web_server_port(7878) #Specify a particular port, as it will be used in the redirect.
  oauth.start_web_server() #In most cases, the webserver is automatically started when get_authorization is called.  In this case, due to the use of oauth.nsoftware.com, it is started manually.
  oauth.set_return_url("https://oauth.nsoftware.com/oauthdemo") #This URL should NOT be used in a production application.  It is only for testing purposes.
  oauth.add_param("state", "http://localhost:" + str(oauth.get_web_server_port())) #The oauth.nsoftware.com server uses this value to redirect to the local server embedded within the OAuth class.
  
  #oauth.set_refresh_token("") #Set to value from previous runs to expedite the login process.
  authStr = oauth.get_authorization()
  #print(oauth.get_refresh_token()) #refresh_token may be used to automatically obtain a new token without user interaction.

def printMenu():
  print(
  """
  Type ? to view this menu, or type one of the commands below to perform the corresponding operation:
	ls    - List Items
	cd    - Change Folders
	get   - Download Items
	put   - Upload Items
	del   - Delete Items
	mkdir - Create Folder
	req   - Create Request Link
	shr   - Create Share Link
	q     - Quit
  """)

def displayList(folderId):
  global authStr, subdomain, currentFolderId, currentParentId, sharefile
  sharefile.set_account_subdomain(subdomain)
  sharefile.set_authorization(authStr)

  sharefile.get_item_info(folderId)
  currentFolderId = sharefile.get_item_id(0)
  currentParentId = sharefile.get_item_parent_id(0)
  frmStr = "%-3.2s%-40.38s%-10.8s%-25.23s"
  print(frmStr % ("#","Title","Type","Creation Date"))
  print(frmStr % ("-1","..","Folder","")) #Parent Directory

  sharefile.list_items(currentFolderId)
  for i in range(sharefile.item_count):
    print( frmStr % (str(i), sharefile.get_item_name(i), resolveItemType(sharefile.get_item_type(i)) 
      ,sharefile.get_item_creation_date(i)))
  return

def resolveItemType(type):
  if type == 0:
    return "File"
  elif type == 1:
    return "Folder"
  elif type == 2:
    return "Link"
  elif type == 3:
    return "Note"
  elif type == 4:
    return "Symbolic Link"
  else:
    return "Unknown"


def prompt(prompt, default):
  res = ""
  res = input(prompt + " [" + default + "]: ")
  if res == "":
    res = default
  return res
  
def main():
  global subdomain, currentFolderId, currentParentId, sharefile
  print("*****************************************************************")
  print("* This demo shows how to use the ShareFile component to list,   *")
  print("* upload, download, delete, and share documents from ShareFile. *")
  print("* Default prompt values are in [ ] where applicable.            *")
  print("*****************************************************************")
  subdomain = input("Account Subdomain: ")

  authenticate()

  displayList("home")
  printMenu()
  response = ""
  while not response.startswith('q') :
    response = input(">")
    if response == "ls":
      displayList(currentFolderId)
    elif response == "cd":
      cd_res = prompt("Number corresponding to folder you wish to view", str(-1))
      if int(cd_res) < 0:
        displayList(currentParentId)
      elif sharefile.get_item_type(int(cd_res)) == 1 :
        displayList(sharefile.get_item_id(int(cd_res)))
      else:
        print("Please choose a valid number corresponding to a folder.")
    elif response == "get":
      get_num_res = input("Number corresponding to item you wish to download: ")
      get_dir_res = prompt("Directory in which to download", "./")
      if not os.path.isdir(get_dir_res):
        print("Please choose a valid directory.")
        break
      filename = get_dir_res + sharefile.get_item_name(int(get_num_res))
      if sharefile.get_item_type(int(get_num_res)) == 1:
        filename += ".zip"
      sharefile.set_local_file(filename)
      sharefile.download_file(sharefile.get_item_id(int(get_num_res)))
      print("File downloaded to: " + filename)
    elif response == "put":
      put_file_res = input("Path to file that will be uploaded: ")
      if not os.path.isfile(put_file_res):
        print("Please input valid file.")
        break
      sharefile.set_local_file(put_file_res)
      sharefile.upload_file(os.path.basename(put_file_res), currentFolderId)
      print("File uploaded with name " + os.path.basename(put_file_res))
      displayList(currentFolderId)
    elif response == "del":
      del_num_res = input("Number corresponding to item you wish to delete: ")
      sharefile.delete_item(sharefile.get_item_id(int(del_num_res)))
      displayList(currentFolderId)
    elif response == "mkdir":
      mkdir_name_res = input("Name of new folder: ")
      sharefile.create_folder(mkdir_name_res, currentFolderId)
      displayList(currentFolderId)
    elif response == "req":
      print(sharefile.create_request_link(currentFolderId))
    elif response == "shr":
      shr_num_res = input("Number corresponding to item you wish to share: ")
      print(sharefile.create_link(sharefile.get_item_id(int(shr_num_res))))
    elif response == "q":
      exit()
    else:
      printMenu()

main()

